File: Deep/Deep/Features/Onboarding/Components/CraftingChecklistRow.swift
import SwiftUI

/// One line in the "crafting your space" loader. Ticks from a hollow ring to a
/// filled lavender check as each step completes, blooming in gently. Mirrors the
/// Calm reference's "We're crafting your sleep plan" checklist, re-voiced.
struct CraftingChecklistRow: View {
  private let label: Text
  private let isDone: Bool

  /// For a caller whose title is chosen at runtime. Such a title does not pass
  /// through the String Catalog.
  init(title: String, isDone: Bool) {
    self.label = Text(title)
    self.isDone = isDone
  }

  /// For a literal title, which resolves through the String Catalog against the
  /// view's locale — how the paywall's benefit list reaches Thai.
  init(_ titleKey: LocalizedStringKey, isDone: Bool) {
    self.label = Text(titleKey)
    self.isDone = isDone
  }

  var body: some View {
    HStack(spacing: 14) {
      ZStack {
        if isDone {
          Image(systemName: "checkmark.circle.fill")
            .font(.system(size: 22))
            .foregroundStyle(.lavenderMist)
            .transition(.scale(scale: 0.6).combined(with: .opacity))
        } else {
          Circle()
            .strokeBorder(Color.driftGrey.opacity(0.4), lineWidth: 1.5)
            .frame(width: 22, height: 22)
        }
      }

      label
        .font(DeepType.body)
        .foregroundStyle(isDone ? .deepPlum : .driftGrey)

      Spacer(minLength: 0)
    }
    .animation(.bloom, value: isDone)
    .accessibilityElement(children: .combine)
    .accessibilityValue(isDone ? "done" : "in progress")
  }
}

#Preview("Crafting checklist row") {
  ZStack {
    AtmosphereBackground()
    VStack(alignment: .leading, spacing: .rhythm) {
      CraftingChecklistRow(title: "Gathering a little calm", isDone: true)
      CraftingChecklistRow(title: "Listening to what you shared", isDone: true)
      CraftingChecklistRow(title: "Shaping your space", isDone: false)
    }
    .padding(.horizontal, .edge)
  }
}

File: android/app/src/main/java/io/appbeyond/freelance/deep/feature/onboarding/components/CraftingChecklistRow.kt
package io.appbeyond.freelance.deep.feature.onboarding.components

import androidx.compose.animation.AnimatedContent
import androidx.compose.animation.animateColorAsState
import androidx.compose.animation.fadeIn
import androidx.compose.animation.fadeOut
import androidx.compose.animation.scaleIn
import androidx.compose.animation.togetherWith
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.semantics.stateDescription
import androidx.compose.ui.tooling.preview.Preview
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import io.appbeyond.freelance.deep.R
import io.appbeyond.freelance.deep.theme.DeepType
import io.appbeyond.freelance.deep.theme.bloom
import io.appbeyond.freelance.deep.theme.deepPlum
import io.appbeyond.freelance.deep.theme.driftGrey
import io.appbeyond.freelance.deep.theme.edge
import io.appbeyond.freelance.deep.theme.rhythm

private val ROW_SPACING = 14.dp
private const val RING_ALPHA = 0.4f

/**
 * One line of the "shaping your space" loader: a hollow ring that blooms into
 * a lavender check as its step completes, the title brightening from drift
 * grey to plum with it.
 *
 * Ported from Deep/Deep/Features/Onboarding/Components/CraftingChecklistRow.swift.
 * TalkBack reads the row as one element with a done / in-progress state.
 */
@Composable
fun CraftingChecklistRow(
  title: String,
  isDone: Boolean,
  modifier: Modifier = Modifier,
) {
  val textColor by animateColorAsState(
    targetValue = if (isDone) Color.deepPlum else Color.driftGrey,
    animationSpec = bloom(),
    label = "crafting-row-text",
  )
  val state = stringResource(if (isDone) R.string.onboarding_step_done else R.string.onboarding_step_in_progress)

  Row(
    modifier = modifier
      .fillMaxWidth()
      .semantics(mergeDescendants = true) { stateDescription = state },
    horizontalArrangement = Arrangement.spacedBy(ROW_SPACING),
    verticalAlignment = Alignment.CenterVertically,
  ) {
    AnimatedContent(
      targetState = isDone,
      transitionSpec = {
        (scaleIn(bloom(), initialScale = MARK_INITIAL_SCALE) + fadeIn(bloom())) togetherWith fadeOut(bloom())
      },
      label = "crafting-row-mark",
    ) { done ->
      if (done) FilledCheckMark() else HollowMark(alpha = RING_ALPHA)
    }

    Text(text = title, style = DeepType.body, color = textColor)
  }
}

@Preview(showBackground = true, name = "Crafting checklist row")
@Composable
private fun CraftingChecklistRowPreview() {
  OnboardingPreviewBackdrop {
    Column(
      modifier = Modifier.padding(horizontal = Dp.edge, vertical = Dp.rhythm),
      verticalArrangement = Arrangement.spacedBy(Dp.rhythm),
    ) {
      CraftingChecklistRow(title = "Gathering a little calm", isDone = true)
      CraftingChecklistRow(title = "Listening to what you shared", isDone = true)
      CraftingChecklistRow(title = "Shaping your space", isDone = false)
    }
  }
}
